### PR TITLE
DEVOPS-6019: Load data from APIs on demand

### PR DESCRIPTION
**Pull request by rallyben**
[DEVOPS-6019](https://jira.rallyhealth.com/browse/DEVOPS-6019)

Lazy loading provides more up to data results.  The previous eager
loading showed data inconsistencies.  Also added logic to only relocate
one task per application per cycle.

### CODE REVIEW

File: carousel/handlers/marathon.py [hunk 1]
```diff
@@ -29,11 +31,15 @@ def get_apps_data(self):
         assert request.status_code == 200
         return request.json()['apps']
 
-    def def_app_data(self, app_id: str):
+    def get_app_data(self, app_id: str):
         '''Retreive data from a single Marathon applcation'''
         url = self.base_url + '/v2/apps' + app_id
         request = self.session.get(url, json={})
-        assert request.status_code == 200
+        if request.status_code != 200:
```
Comment (bazimov) [inline carousel/handlers/marathon.py]: You could use requests built in raise_for_status()
http://docs.python-requests.org/en/master/user/quickstart/#response-status-codes

```
>>> bad_r = requests.get('http://httpbin.org/status/404')
>>> bad_r.status_code
404

>>> bad_r.raise_for_status()
Traceback (most recent call last):
  File "requests/models.py", line 832, in raise_for_status
    raise http_error
requests.exceptions.HTTPError: 404 Client Error
```

Comment (bazimov) [reply carousel/handlers/marathon.py]: If you want to change it later after this commit, I am fine with that too. I think all asserts on requests status code can be replaced with raise_for_status().

**Comment by rallyben** [comment]
```
$ HTTP_PROXY=socks5://localhost:3130 AWS_PROFILE=rally-dev AWS_DEFAULT_REGION=us-east-1 ./blackout.py --marathon http://mesos-master.mesos-infra.werally.plumbing:8080 --mesos http://mesos-master.mesos-infra.werally.plumbing:5050 --asg mesos-slave-asg-mesos-infra
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state Omega: All instances up-to-date
(.virtual) ben:carousel benheilman$ HTTP_PROXY=socks5://localhost:3130 AWS_PROFILE=rally-dev AWS_DEFAULT_REGION=us-east-1 ./blackout.py --marathon http://mesos-master.mesos-infra.werally.plumbing:8080 --mesos http://mesos-master.mesos-infra.werally.plumbing:5050 --asg mesos-slave-asg-mesos-infra
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state Alpha: Target instance i-09decfe65597762f1 is not yet tagged
INFO:Roller:Applying tag to instance
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state 1: Target instance i-09decfe65597762f1 is still attached to ASG
INFO:Roller:Detaching instance
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state 2: Target instance i-09decfe65597762f1 is hosting tasks
INFO:Roller:Draining instance
INFO:Mesos:Reserving 0.30 cpus and 213 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Marathon:Restarting application /dev/experiments/hello-play-two
INFO:root:Refusing to relocate second task for an application dev_experiments_hello-play-two.fc37b6ef-7eff-11e8-902d-12f9cbd9fd34
INFO:Marathon:Restarting application /dev-zerocats/sortinghat/web
INFO:Marathon:Restarting application /integration/mayo/mayo-ui
INFO:Marathon:Killing task mesos-infra_marathon-lb_marathon-lb-internal.87af6893-7eff-11e8-902d-12f9cbd9fd34
INFO:Marathon:Restarting application /dev-zerocats/experiments/hello-play
INFO:Marathon:Killing task dev_experiments_hello-play-one.a3e9aa75-7eff-11e8-902d-12f9cbd9fd34
INFO:root:Refusing to relocate second task for an application dev_experiments_hello-play-one.a3e9f896-7eff-11e8-902d-12f9cbd9fd34
INFO:Mesos:Reserving 0.30 cpus and 1536 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.20 cpus and 64 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.10 cpus and 50 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.10 cpus and 2560 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.10 cpus and 512 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.10 cpus and 512 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.10 cpus and 512 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state 2: Target instance i-09decfe65597762f1 is hosting tasks
INFO:Roller:Draining instance
INFO:Marathon:Killing task dev_experiments_hello-play-one.a3e9f896-7eff-11e8-902d-12f9cbd9fd34
INFO:Mesos:Reserving 0.10 cpus and 512 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S40
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state 3: Target instance i-09decfe65597762f1 is drained
INFO:Roller:Terminating instance
INFO:aws:Terminating instance i-09decfe65597762f1
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state Alpha: Target instance i-05c04b4cc90988adf is not yet tagged
INFO:Roller:Applying tag to instance
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state 1: Target instance i-05c04b4cc90988adf is still attached to ASG
INFO:Roller:Detaching instance
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state 2: Target instance i-05c04b4cc90988adf is hosting tasks
INFO:Roller:Draining instance
INFO:Mesos:Reserving 0.30 cpus and 213 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Marathon:Restarting application /dev-zerocats/experiments/hello-play
INFO:Marathon:Restarting application /dev/experiments/hello-play-two
INFO:root:Refusing to relocate second task for an application dev_experiments_hello-play-two.86b932ba-7f08-11e8-902d-12f9cbd9fd34
INFO:Marathon:Restarting application /integration/mayo/mayo-ui
INFO:Marathon:Restarting application /dev-zerocats/sortinghat/web
INFO:Marathon:Killing task dev_experiments_hello-play-one.d4f6bbdc-7eff-11e8-902d-12f9cbd9fd34
INFO:Marathon:Killing task mesos-infra_marathon-lb_marathon-lb-internal.d56f202d-7eff-11e8-902d-12f9cbd9fd34
INFO:root:Refusing to relocate second task for an application dev_experiments_hello-play-one.631729f1-7f00-11e8-902d-12f9cbd9fd34
INFO:Mesos:Reserving 0.30 cpus and 1536 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.20 cpus and 64 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.10 cpus and 50 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.10 cpus and 2560 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.10 cpus and 512 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.10 cpus and 512 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.10 cpus and 512 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state 2: Target instance i-05c04b4cc90988adf is hosting tasks
INFO:Roller:Draining instance
INFO:Marathon:Killing task dev_experiments_hello-play-one.631729f1-7f00-11e8-902d-12f9cbd9fd34
INFO:Mesos:Reserving 0.10 cpus and 512 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:Mesos:Reserving 0.10 cpus and 32 mem on 16fd6e06-7376-4cfe-80ad-8f8ca3d2806e-S41
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state 3: Target instance i-05c04b4cc90988adf is drained
INFO:Roller:Terminating instance
INFO:aws:Terminating instance i-05c04b4cc90988adf
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.credentials:Found credentials in shared credentials file: ~/.aws/credentials
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
INFO:Roller:Detected state Omega: All instances up-to-date
```

File: carousel/handlers/marathon.py (hunk 1 above)
Comment (rallyben) [reply carousel/handlers/marathon.py]: Good eye. I'll get that on the next PR

**Comment by TimGebhardt** [comment]
FYI PR #1 didn't include a Readme file.  Wanna add that to the next PR please?

**Pull request merged**